Source/Core/Duality/Utility/Coroutines/Coroutine.cs

```diff
@@ -0,0 +1,134 @@
+using System.Linq;
+using Duality.Resources;
+
+namespace Duality
+{
+	public sealed class Coroutine : IDisposable
+	{
+		/// <summary>
+		/// Starts a coroutine and registers it in the scene's CoroutineManager
+		/// </summary>
+		/// <param name="scene">The scene that will execute the coroutine</param>
+		/// <param name="method">The coroutine implementation</param>
+		/// <param name="name">The name of the coroutine, optional</param>
+		/// <returns>The created corutine</returns>
+		public static Coroutine Start(Scene scene, IEnumerable<IWaitCondition> method, string name = null)
+		{
+			return Start(scene, method.GetEnumerator(), name);
+		}
+
+		/// <summary>
+		/// Starts a coroutine and registers it in the scene's CoroutineManager
+		/// </summary>
+		/// <param name="scene">The scene that will execute the coroutine</param>
+		/// <param name="enumerator">The coroutine IWaitConditions enumerator</param>
+		/// <param name="name">The name of the coroutine, optional</param>
+		/// <returns>The created corutine</returns>
+		public static Coroutine Start(Scene scene, IEnumerator<IWaitCondition> enumerator, string name = null)
+		{
+			return scene.CoroutineManager.StartNew(enumerator, name);
+		}
+
+		private IEnumerator<IWaitCondition> enumerator = null;
+
+		public CoroutineStatus Status { get; private set; }
+
+		public Exception LastException { get; private set; }
+
+		public string Name { get; private set; }
+
+		public object Tag { get; private set; }
+
+		public bool IsAlive
+		{
+			get { return this.Status == CoroutineStatus.Paused || this.Status == CoroutineStatus.Running; }
+		}
+
+		internal void Setup(IEnumerator<IWaitCondition> values, string name)
+		{
+			this.Status = CoroutineStatus.Running;
+
+			this.LastException = null;
+			this.Name = name;
+
+			if (this.enumerator != null)
+				this.enumerator.Dispose();
+
+			this.enumerator = values;
+
+			if (!this.enumerator.MoveNext())
+				this.Status = CoroutineStatus.Complete;
+		}
+
+		/// <summary>
+		/// Updates the coroutine's current IWaitCondition and executes the coroutine's code until it finds the next one.
+		/// </summary>
+		/// <remarks>
+		/// A null currentYield is ignored
+		/// </remarks>
+		internal void Update()
+		{
+			if (this.Status != CoroutineStatus.Running) return;
+
+			try
+			{
+				IWaitCondition currentYield = this.enumerator.Current;
+				if (currentYield == null || currentYield.Update())
+				{
+					if (!this.enumerator.MoveNext())
+						this.Status = CoroutineStatus.Complete;
+				}
+			}
+			catch(Exception e)
+			{
+				Logs.Core.WriteError("An error occurred while processing Coroutine '{0}': {1}", this.Name, LogFormat.Exception(e));
+
+				this.LastException = e;
+				this.Status = CoroutineStatus.Error;
+			}
+		}
+		/// <summary>
+		/// Restarts the coroutine by moving the enumerator back to the first element.
+		/// </summary>
+		internal void Restart()
+		{
+			this.enumerator.Reset();
+			this.enumerator.MoveNext();
```

Comment:
Only Reset should be called here.

The Update method should be changed so that it always first calls IEnumerable.MoveNext before calling IEnumerable.Current.

This is consistent with how foreach works.

Reply:
Well, apparently IEnumerator.Reset() now throws a NotSupportedException. Can it be useful to restart a Coroutine?

Reply:
Hmm seems Reset is not supposed to be used but to reset a enumerable you have to call GetEnumerator again.

Reply:
> The Update method should be changed so that it always first calls IEnumerable.MoveNext before calling IEnumerable.Current.

It's made like this on purpose, because it needs to be able to re-evaluate the Current `IWaitCondition` on each frame until it's satisfied (or go ahead when null), hence why there was a first `MoveNext()` inside Setup and Restart methods.

Reply:
Wouldnt it just work like this if remove the movenext from Setup?
```
		internal void Setup(IEnumerator<IWaitCondition> values, string name)
		{
			this.Status = CoroutineStatus.Running;

			this.LastException = null;
			this.Name = name;

			if (this.enumerator != null)
				this.enumerator.Dispose();

			this.enumerator = values;
		}

		internal void Update()
		{
			if (this.Status != CoroutineStatus.Running) return;

			try
			{
				if (!this.enumerator.MoveNext())
					this.Status = CoroutineStatus.Complete;

				this.enumerator.Current?.Update();
			}
			catch(Exception e)
			{
				Logs.Core.WriteError("An error occurred while processing Coroutine '{0}': {1}", this.Name, LogFormat.Exception(e));

				this.LastException = e;
				this.Status = CoroutineStatus.Error;
			}
		}
```

Reply:
No, because the MoveNext would be called first thing on each Update cycle, and I want it to stay there until the Current IWaitCondition is fulfilled (i.e. this.enumerator.Current.Update() returns true - or is null).

I'll check about the Reset part

Reply:
Ah I see you are right 👍